src/privileged/autoplay/api.js

```diff
@@ -0,0 +1,365 @@
+            }
+            callback({
+              timestamp: Date.now(),
+              globalSettings: {
+                allowAutoPlay: autoplayState,
+              },
+            });
+          };
+
+          Services.obs.addObserver(pageSettingObs, "perm-changed");
+          Preferences.observe("media.autoplay.default", globalSettingObs);
+
+          return () => {
+            Services.obs.removeObserver(pageSettingObs, "perm-changed");
+            Preferences.ignore("media.autoplay.default", globalSettingObs);
+          };
+        }).api(),
+
+        audibleAutoplayOccurred: new EventManager(context, "autoplay.audibleAutoplayOccurred", fire => {
+          const autoplayObs = (subject, topic, data) => {
+            fire.async(tabManager.getWrapper(subject).id,
+                       subject.linkedBrowser.currentURI.spec);
+          };
+          Services.obs.addObserver(autoplayObs, "AudibleAutoplayMediaOccurred");
+
+          return () => {
+            Services.obs.removeObserver(autoplayObs, "AudibleAutoplayMediaOccurred");
+          };
+        }).api(),
+
+        setPreferences: async(variation) => {
+          this.branch = variation;
+
+          let cacheDefaultSetting =
+            await ExtensionStorage.get(extension.id, "media.autoplay.default").catch(Cu.reportError);
+          if (!cacheDefaultSetting.hasOwnProperty("media.autoplay.default")) {
+            // Save current preferences setting and recover preferences when study ends.
+            const defaultSetting = {
+              "media.autoplay.default" : Preferences.get("media.autoplay.default"),
+              "media.autoplay.enabled.user-gestures-needed": Preferences.get("media.autoplay.enabled.user-gestures-needed"),
+              "media.autoplay.ask-permission": Preferences.get("media.autoplay.ask-permission"),
+            }
+            await ExtensionStorage.set(extension.id, defaultSetting).catch(Cu.reportError);
+          }
+
+          Preferences.set({
+            "media.autoplay.default": variation === "control" ? 0 : 2,
+            "media.autoplay.enabled.user-gestures-needed": true,
+            "media.autoplay.ask-permission": true,
+          });
+          setAutoplayPromptLayout(variation);
+        },
+
+        clearPreferences: async() => {
+          function getDefaultPreference(pref) {
+            return ExtensionStorage.get(extension.id, pref).catch(Cu.reportError);
+          }
+
+          const defaultSetting = [
+            await getDefaultPreference("media.autoplay.default"),
+            await getDefaultPreference("media.autoplay.enabled.user-gestures-needed"),
+            await getDefaultPreference("media.autoplay.ask-permission"),
+          ];
+
+          let setting = {};
+          for (let item of defaultSetting) {
+            for (let key in item) {
+              setting[key] = item[key];
+            }
+          }
+          Preferences.set(setting);
```

Comment:
You can just use `Preferences.reset` instead, no?

https://searchfox.org/mozilla-central/rev/3fa761ade83ed0b8ab463acb057c2cf0b104689e/toolkit/modules/Preferences.jsm#199

Reply:
@johannhof But I think `Preferences.reset` would clear all the user-set value to default, instead of changing value to the value which users have before starting shield study, right?